schema/common.json

```diff
@@ -4,6 +4,103 @@
   "description": "Serverless Workflow specification - common schema",
   "type": "object",
   "definitions": {
+    "expressionMarkup": {
+      "type": "string",
+      "pattern": "^\\${.*}$"
+    },
+    "jsonpointer": {
+      "$ref": "#/definitions/expressionMarkup"
+    },
+    "functionMarkup": {
+      "type": "string",
+      "pattern": "^\\$fn{.*}$"
+    },
+    "markup": {
+      "oneOf": [
+        {
+          "$ref": "#/definitions/expressionMarkup"
+        },
+        {
+          "$ref": "#/definitions/functionMarkup"
+        }
+      ]
+    },
+    "booleanOrMarkup": {
+      "oneOf": [
+        {
+          "$ref": "#/definitions/markup"
+        },
+        {
+          "type": "boolean"
+        }
+      ]
+    },
+    "numberOrMarkup": {
+      "oneOf": [
+        {
+          "$ref": "#/definitions/markup"
+        },
+        {
+          "type": "number"
+        }
+      ]
+    },
+    "stringOrMarkup": {
+      "oneOf": [
+        {
+          "$ref": "#/definitions/markup"
+        },
+        {
+          "type": "string",
+          "pattern": "^(?!\\$)"
+        }
+      ]
+    },
+    "arrayOrMarkup": {
+      "oneOf": [
+        {
+          "$ref": "#/definitions/markup"
+        },
+        {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/anyOrMarkup"
+          }          
+        }
+      ]
+    },
+    "objectOrMarkup": {
+      "oneOf": [
+        {
+          "$ref": "#/definitions/markup"
+        },
+        {
+          "type": "object",
+          "additionalProperties": {
+            "$ref": "#/definitions/anyOrMarkup"
+          }
+        }
+      ]
+    },
+    "anyOrMarkup": {
```

Comment:
I think that this is here only for runtimes, not the spec
there is no reason imo to define this inside DSL where runtimes can determine the result type from the expression alone.

Reply:
This is not about the result type of the expression. It's about declaring fields that can also be defined using an expression. Let's say we want to define a parameter that's supposed to be a number, e.g. number of repeats. If you also want to allow an expression, it can be either a string with format `${ }` or it can be a number, but it can neither be a random string nor an object. This is about getting the workflow definition parsed with useful types.